File: 第一章/learning_action/scripts/count_server.py
#!/usr/bin/env python
# coding:utf-8

import rospy
import actionlib
from learning_action.msg import CountAction, CountGoal, CountFeedback, CountResult

class CountActionServer :
    goal = CountGoal()

    def __init__(self) :
        self.server = actionlib.SimpleActionServer("count", 
        CountAction, self.ExecuteCb, False)
        self.server.register_preempt_callback(self.PreemptCb)

    def Start(self) :
        self.server.start()

    def ExecuteCb(self, goal) :
        rate =rospy.Rate(1)
        rospy.loginfo("Get the goal,the goal value is: %d", goal.goal_num)
        feedback = CountFeedback()
        result = CountResult()
        count_num = 0
        finish_num = 10
        while count_num < finish_num :
            count_num += 1
            rospy.loginfo("Counting the number: %d",count_num)
            feedback.percent_complete = (float(count_num)/float(finish_num))*100
            self.server.publish_feedback(feedback)
            rate.sleep()
            result.finish = count_num
        if (self.server.is_active()) :
            self.server.set_succeeded(result)

    def PreemptCb(self) :
        if(self.server.is_active()) :
            self.server.set_preempted()

if __name__ == '__main__' :
    rospy.init_node("count_server")
    countserver = CountActionServer()
    countserver.Start()
    rospy.spin()

File: chapter 2/learning_action/src/count_server.cpp
#include "ros/ros.h"
#include "actionlib/server/simple_action_server.h"
#include "learning_action/CountAction.h"

typedef actionlib::SimpleActionServer<learning_action::CountAction> Server;

class CountActionServer
{
  private:
  learning_action::CountFeedback feedback;
  learning_action::CountResult   result;
  public:
  Server server;
  CountActionServer(ros::NodeHandle nh):
  server(nh,"count",boost::bind(&CountActionServer::ExecuteCb,this,_1),false)
  {
    server.registerPreemptCallback(boost::bind(&CountActionServer::preemptCb,this));
  }

  void Start()
  {
    server.start();
  }

  void ExecuteCb(const learning_action::CountGoalConstPtr& goal)
  {
    ros::Rate rate(1);
    ROS_INFO("Get the goal,the goal value is: %d", goal->goal_num);
    // learning_action::CountFeedback feedback;
    // learning_action::CountResult   result;
    int count_num  = 1;
    int finish_num = 10;
    for (; count_num <= finish_num; count_num++)
    {
      ROS_INFO("Counting the number: %d",count_num);
      feedback.percent_complete = (float)count_num/finish_num*100;
      server.publishFeedback(feedback);
      rate.sleep();
      result.finish = count_num;

    }
    // result.finish = count_num;

    if(server.isActive())
    server.setSucceeded(result);
    
  }

  void preemptCb()
  {
    if(server.isActive())
    server.setPreempted();
  }
};

int main(int argc, char **argv)
{
  ros::init(argc, argv, "count_server");
  ros::NodeHandle nh;
  CountActionServer countserver(nh);
  countserver.Start();
  ros::spin();
  return 0;
}

